### PR TITLE
MAINT: Translate x86-qsort.dispatch.c.src to C++ using templates.

### PR DESCRIPTION
**Pull request by serge-sans-paille**
Using a approach close to engine like xsimd here, to abstract the vectorized
code.

### CODE REVIEW

**Comment by seberg** [comment]
Have not looked at it, but could you make sure to rename/`git mv` the file rather than delete and create a new one again?

**Comment by serge-sans-paille** [comment]
> Have not looked at it, but could you make sure to rename/`git mv` the file rather than delete and create a new one again?

I Moved it, then modified it to do the conversion. Split that in two steps.

**Comment by seberg** [comment]
Ah, thanks.  Is github just weird about showing moved files only well if the diffs are small/moved without change?

**Comment by serge-sans-paille** [comment]
green and ready for review

**Comment by serge-sans-paille** [comment]
ping :-) This one was a bit more difficult to handle than the others, but thanks to the (thin) simd abstraction, everything went well.

File: numpy/distutils/ccompiler_opt.py [hunk 1]
```diff
@@ -2551,6 +2551,8 @@ def _generate_config(self, output_dir, dispatch_src, targets, has_baseline=False
         except OSError:
             pass
 
+        os.makedirs(os.path.dirname(config_path), exist_ok=True)
+
```
Comment (mattip) [inline numpy/distutils/ccompiler_opt.py:2555]: Any idea why this is needed now?

File: numpy/core/src/common/simd/avx512/reorder.h [hunk 2]
```diff
@@ -214,13 +214,13 @@ NPY_FINLINE npyv_u16 npyv_rev64_u16(npyv_u16 a)
 
 NPY_FINLINE npyv_u32 npyv_rev64_u32(npyv_u32 a)
```
Comment (mattip) [inline numpy/core/src/common/simd/avx512/reorder.h:215]: Was this (and `npyv_rev64_f32`) ever called?

**Comment by r-devulap** [comment]
> ping :-) This one was a bit more difficult to handle than the others, but thanks to the (thin) simd abstraction, everything went well.

The patch looks great! quick question w.r.t the SIMD abstraction: does it in anyway overlap with abstracting universal intrinsics using C++ templates? I am not sure what the scope of that work is, so it might not be a relevant question at all :)

File: numpy/distutils/ccompiler_opt.py (hunk 1 above)
Comment (serge-sans-paille) [reply numpy/distutils/ccompiler_opt.py:2555]: yeah, it happens when there's no `.c.src` remaining in a setup.py subdirectory, but the system somehow needs that subdirectory to be created.

File: numpy/core/src/common/simd/avx512/reorder.h (hunk 2 above)
Comment (serge-sans-paille) [reply numpy/core/src/common/simd/avx512/reorder.h:215]: I don't know, I'm just a translator ;-)

**Comment by serge-sans-paille** [comment]
> > ping :-) This one was a bit more difficult to handle than the others, but thanks to the (thin) simd abstraction, everything went well.
> 
> The patch looks great! quick question w.r.t the SIMD abstraction: does it in anyway overlap with abstracting universal intrinsics using C++ templates? I am not sure what the scope of that work is, so it might not be a relevant question at all :)

At some point I'd love to use the xsimd abstraction instead, but that's more of a midterm goal, so the slight abstraction here is fine (to me).

**Comment by serge-sans-paille** [comment]
It's just a translation from C.src to C++, I'm not adding (nor removing !) architecture support etc, in the spirit of the previous PR on the *road to C++*. In a similar manner, I'm not changing the original algorithm and don't plan to (well, that's a great topic and I'd love investing time on it, but that's another story, and not this PR's).

Concerning the genericity, using a generic simd abstraction (like the one provided by xsimd) indeed looks beneficial ;-)

**Comment by charris** [comment]
Thanks @serge-sans-paille.

Yes, the questions involve the original. I'd like to see more on that somewhere along the line. There is also the problem of indirect sorts.

**Comment by r-devulap** [comment]
@charris The speed up comes really from three sources: 

1) Vectorizing the quicksort partitioning algorithm. AVX-512 _compressstore_ instruction comes in super handy for partitioning and provides a very nice speed up. 
2) Bitonic sorting networks for smaller arrays (<= 128 elements). It parallelizes a lot of the compare and swap. Additionally, the fact that you perform the compare swap operations in the AVX-512 registers means you read and write to memory a lot fewer times. 
3) Picking the median of randomly chosen 72 elements. This isn't as beneficial as the previous two, but still adds value specially when the distribution is uniformly random. 

Hope this helps.

**Comment by charris** [comment]
> AVX-512 compressstore instruction comes in super handy for partitioning

I'd like to understand how that works.

Can any of this be generalized for 1) indirect sorts, 2) merging, as in plain old mergesort. I'm more interested in the possibility of having some useful general routines. What would also interesting is some curves of `time/(n*log(n))` vs `log(n)` in an attempt to tease out the value of the different parts. Maybe even some profiling. Is any of that in the original papers? 

What bothers me about the current code is its complexity and restriction to 32 bits on avx512, and perhaps the lack of indirect sorts.

**Pull request merged**